File: iterator/java/IteratorPatternDemo.java
interface Iterator {
	boolean hasNext();
	<T> T next();
}

interface Collection {

	Iterator getIterator();

	Iterator getIteratorReversed();
}

final class NameRepository implements Collection {

	private final String[] langs;

	NameRepository() {
		this.langs = new String[]{"Java", "Kotlin", "C++", "Python", "Rust", "C#", "Ruby", "Pearl", "JS"};
	}

	@Override
	public Iterator getIterator() {
		return new Iterator() {

			private int index;

			@Override
			public boolean hasNext() {
				return index >= 0 && index < langs.length;
			}

			@Override
			public <T> T next() {
				return (T) langs[index++];
			}
		};
	}

	@Override
	public Iterator getIteratorReversed() {
		return new Iterator() {

			private int index = langs.length - 1;
			
			@Override
			public boolean hasNext() {
				return index >= 0 && index < langs.length;
			}

			@Override
			public <T> T next() {
				return (T) langs[index--];
			}
		};
	}
}

final class IteratorPatternDemo {

	public static void print(Iterator iter) {
		while (iter.hasNext())
			System.out.print(iter.<String>next()+" ");
		System.out.println();
	}

	public static void main(String[] args) {
		NameRepository nameRepository = new NameRepository();
		Iterator iter = nameRepository.getIterator();
		print(iter);

		iter = nameRepository.getIteratorReversed();
		print(iter);
	}
}

File: iterator/kotlin/iterator_pattern_demo.kt
interface Iterator {
	fun hasNext(): Boolean;
	fun <T> next(): T;
}

interface Collection {

	fun getIterator(): Iterator;

	fun getIteratorReversed(): Iterator;
}

class NameRepository(private val langs: Array<String>) : Collection {

	constructor() : this(arrayOf<String>("Java", "Kotlin", "C++", "Python", "Rust", "C#", "Ruby", "Pearl", "JS"))

	override fun getIterator(): Iterator {
		return object : Iterator {

			private var index: Int = 0;
	
			override fun hasNext(): Boolean {
				return index >= 0 && index < langs.size;
			}

	
			override fun <T> next(): T {
				return langs[index++] as T;
			}
		};
	}

	override fun getIteratorReversed(): Iterator {
		return object : Iterator {

			private var index: Int = langs.size - 1;
			
			override fun hasNext(): Boolean {
				return index >= 0 && index < langs.size;
			}

	
			override fun <T> next(): T {
				return langs[index--] as T;
			}
		};
	}
}

private fun print(iter: Iterator): Unit {
	while (iter.hasNext())
		print(iter.next<String>()+" ")
	println()
}

fun main(): Unit {
	val collection: Collection = NameRepository()
	var iterator: Iterator = collection.getIterator()
	print(iterator)

	iterator = collection.getIteratorReversed()
	print(iterator)
}